r-package/grf/R/regression_forest.R

```diff
@@ -84,100 +87,83 @@
-      mtry = validate_mtry(mtry, X),
-      alpha = validate_alpha(alpha),
-      imbalance.penalty = validate_imbalance_penalty(imbalance.penalty),
-      honesty.fraction = validate_honesty_fraction(honesty.fraction, honesty),
-      honesty.prune.leaves = validate_honesty_prune_leaves(honesty.prune.leaves)
-    )
-  }
+  tunable.params <- c("sample.fraction", "mtry", "min.node.size", "honesty.fraction",
+                      "honesty.prune.leaves", "alpha", "imbalance.penalty")
 
   data <- create_data_matrices(X, outcome = Y, sample.weights = sample.weights)
-  forest <- regression_train(
-    data$train.matrix, data$sparse.train.matrix,
-    data$outcome.index, data$sample.weight.index,
-    data$use.sample.weights,
-    as.numeric(tunable.params["mtry"]),
-    num.trees,
-    as.numeric(tunable.params["min.node.size"]),
-    as.numeric(tunable.params["sample.fraction"]),
-    honesty,
-    as.numeric(tunable.params["honesty.fraction"]),
-    as.numeric(tunable.params["honesty.prune.leaves"]),
-    ci.group.size,
-    as.numeric(tunable.params["alpha"]),
-    as.numeric(tunable.params["imbalance.penalty"]),
-    clusters,
-    samples.per.cluster,
-    compute.oob.predictions,
-    num.threads,
-    seed
-  )
+  args <- list(clusters = clusters,
+              samples.per.cluster = samples.per.cluster,
+              sample.fraction = sample.fraction,
+              mtry = mtry,
+              min.node.size = min.node.size,
+              honesty = honesty,
+              honesty.fraction = honesty.fraction,
+              honesty.prune.leaves = honesty.prune.leaves,
+              alpha = alpha,
+              imbalance.penalty = imbalance.penalty,
+              ci.group.size = ci.group.size,
+              num.threads = num.threads,
+              seed = seed)
+
+  if (tune.parameters[1L] != "none") {
+    if (tune.parameters[1L] == "all") {
+      tune.parameters <- tunable.params
+      if (!honesty) {
+        tune.parameters <- tune.parameters[!grepl("honesty", tune.parameters)]
+      }
+    } else {
+      tune.parameters <- unique(match.arg(tune.parameters, tunable.params, several.ok = TRUE))
+    }
+    tuning.output <- tune_forest(X, data, args, tune.parameters,
+                                forest.type = "regression_forest",
+                                tune.num.trees, tune.num.reps, tune.num.draws)
+
+    args <- modifyList(args, as.list(tuning.output[["params"]]))
+  }
+
+  args[["num.trees"]] <- num.trees
+  args[["compute.oob.predictions"]] <- compute.oob.predictions
+  forest <- do.call.grf(regression_train, c(data, args))
 
   class(forest) <- c("regression_forest", "grf")
   forest[["ci.group.size"]] <- ci.group.size
   forest[["X.orig"]] <- X
   forest[["Y.orig"]] <- Y
   forest[["sample.weights"]] <- sample.weights
   forest[["clusters"]] <- clusters
-  forest[["tunable.params"]] <- tunable.params
-  if (tune.parameters) {
+  forest[["tunable.params"]] <- args[tunable.params]
+  if (exists("tuning.output")) {
```

Comment:
Minor comment: this looks a bit inelegant. How idiomatic is this? Consider just defining `tuning.output` as `NULL` whenever `tune.parameters[1L] == "none"`.